### ISSUE TITLE
X509Chain NotSignatureValid error on Linux with .NET Core 3.0+

### ISSUE THREAD

**Issue by mattw-db** [dotnet/corefx#42717]
Hello,

We're having issues with X509 cert chain validation after retargeting our Linux application to netcoreapp3.0.  The chain in three deep: we have a self-signed root CA, which has signed a domain level CA, which in turn has signed a server (non-CA) cert.

When validating the chain, we're met with a X509ChainStatusFlags.NotSignatureValid error carrying a "certificate signature failure" StatusInformation string.  The error appears to originate from openssl, but the native code responsible for bubbling this up to the managed world seems to be new (or at least heavily reworked) for .NET Core 3.0 and largely introduced by this pull request: https://github.com/dotnet/corefx/pull/35801

The issue occurs only on Linux (both x64 and ARM), not Windows.  It is seen with both .NET Core 3.0 and 3.1, but not with 2.2, 2.1, or 2.0.  I've also tested with both openssl 1.1.1 and 1.0.2 libs and it's seen with both.

I've validated the chain on the command line with both the openssl and GnuTLS certtool command (in the same Linux environments) and both show no errors.

The attached project (with a junk cert chain formed similarly to our own) should reproduce the issue easily.  On Windows (or Linux with .NET Core 2.x) it's output with be "Chain Valid".  On Linux with .NET Core 3.0/3.1, it'll print "Chain Invalid: NotSignatureValid: certificate signature failure"

[CertChainValidateTest.zip](https://github.com/dotnet/corefx/files/3889530/CertChainValidateTest.zip)

Let me know if there's anything else I can provide or should check.

Thanks a lot!

Matt

**Comment by bartonjs** [comment]
The other major change between 2.x and 3.0 is we now prefer loading OpenSSL 1.1.  If you have both 1.0 and 1.1 on your system you can try running with `CLR_OPENSSL_VERSION_OVERRIDE=1.0.0` (or whatever SONAME suffix your 1.0.x has) and see if the error goes away.  If so, it just means there's a more stringent check in the newer version of OpenSSL.

(Just a quick self-diagnosis comment, haven't had time to look at the repro yet)

**Comment by mattw-db** [comment]
Thanks for the tip, but no luck.  I get the same results on a machine with only openssl 1.0.2 installed.

```
mattw@mattw-linuxdev-vb:~/CertChainValidateTest$ openssl version
OpenSSL 1.0.2g  1 Mar 2016
mattw@mattw-linuxdev-vb:~/CertChainValidateTest$ ls -1 /usr/lib/x86_64-linux-gnu/lib{crypto,ssl}*.so | xargs readelf -d | grep SONAME
 0x000000000000000e (SONAME)             Library soname: [libcrypto.so.1.0.0]
 0x000000000000000e (SONAME)             Library soname: [libssl3.so]
 0x000000000000000e (SONAME)             Library soname: [libssl.so.1.0.0]
mattw@mattw-linuxdev-vb:~/CertChainValidateTest$ CLR_OPENSSL_VERSION_OVERRIDE=1.0.0 dotnet run
Chain Invalid: NotSignatureValid: certificate signature failure
```

**Comment by bartonjs** [comment]
Hm.  I can't get it to repro on Ubuntu 16.04 with 3.0.0 (the version pulled down by a build of corefx I did).

```
CertChainValidateTest$ dotnet bin/Debug/netcoreapp3.0/CertChainValidateTest.dll
Chain Valid
```

```
CertChainValidateTest$ dotnet --info
.NET Core SDK (reflecting any global.json):
 Version:   3.0.100
 Commit:    04339c3a26

Runtime Environment:
 OS Name:     ubuntu
 OS Version:  16.04
 OS Platform: Linux
 RID:         ubuntu.16.04-x64
 Base Path:   /home/bartonjs/git/bartonjs/corefx/.dotnet/sdk/3.0.100/

Host (useful for support):
  Version: 3.0.0
  Commit:  7d57652f33

.NET Core SDKs installed:
  3.0.100 [/home/bartonjs/git/bartonjs/corefx/.dotnet/sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.App 3.0.0 [/home/bartonjs/git/bartonjs/corefx/.dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 3.0.0 [/home/bartonjs/git/bartonjs/corefx/.dotnet/shared/Microsoft.NETCore.App]

To install additional .NET Core runtimes or SDKs:
  https://aka.ms/dotnet-download
```

```
CertChainValidateTest$ ls -1 /usr/lib/x86_64-linux-gnu/lib{crypto,ssl}*.so | xargs readelf -d | grep SONAME
 0x000000000000000e (SONAME)             Library soname: [libcrypto.so.1.0.0]
 0x000000000000000e (SONAME)             Library soname: [libssl.so.1.0.0]
 0x000000000000000e (SONAME)             Library soname: [libssl3.so]

```

**Comment by mattw-db** [comment]
@bartonjs,

Thanks for testing.  I think I understand the difference between our environments, though not why it matters.

My issue goes away after deleting my certificate store directory from (~/.dotnet/corefx/cryptography/x509stores/my/)

It appears that "my" store contained a pfx with another old test cert with the same CN that the "domain-ca.cert" in my test program is adding to its X509Certificate2Collection.  The cert in the store is old (actually, from a completely different application) and signed by a different root.

With .NET Core 3.x (not 2.x), the signature validation seems to be ignoring the domain-ca.cert file I'm providing, and instead using the old cert from the store.

As a workaround, deleting thing the offending PFX from my store is fine, but this behavior still seems very odd to me. 

If you want to try to reproduce it, try adding the attached .pfx to ~/.dotnet/corefx/cryptography/x509stores/my/
Just having it there seems to be enough to make the issue come back.

PFX: [pfx.zip](https://github.com/dotnet/corefx/files/3913046/pfx.zip)

**Comment by bartonjs** [comment]
Ah.  It's probably due to the lack of an Authority Key Identifier extension in your certs, and our collection load order changing.

You have two different potential paths when using just IssuerCN->SubjectCN matching:

* server -> provided domain-ca -> provided root-ca.
* server -> store-based domain-ca (-> store-based root-ca (if it's present)).

In 2.1 we searched LM\Root, CU\Root, chainPolicy.ExtraStore, CU\My, CU\CA, LM\CA.  In 3.0 the ExtraStore ended up sliding to last (to massively simplify the disposal logic).  So now it finds the "bad" chain first.  Presumably it's also a problem when the bad options are at the same level in the search space (e.g. both certs in CU\My), except then it's "random" (ordered by something you can't control).

If you had an Authority Key Identifier, then there's extra criteria for determining the eligibility of a candidate CA, either AKID.keyId == candidate.SKID, or AKID.nameAndSerial.name == candidate.SubjectDN && AKId.nameAndSerial.serial == candidate.SerialNumber; which would drop whichever of the bad paths.

I'm not sure how feasible us trying to recover from this state would be... the two easy solutions are: add an AKID on your certs, and/or don't use repeat CA names.


If you're using CertificateRequest to build your certs, you can add an AKID by

```C#
byte[] skid = skidExtension.RawData;
Debug.Assert(skid0] == 0x04);
// Otherwise we need more complex DER building
Debug.Assert(skid.Length <= 127);
byte[] akid = new byte[skid.Length + 2];
akid[0] = 0x30;
akid[1] = (byte)skid.Length;
akid[2] = 0x80;
Buffer.BlockCopy(skid, 1, akid, 3, skid.Length - 1);

request.CertificateExtensions.Add(new X509Extension("2.5.29.35", akid, false));
```

(I'm like 99% sure on that).  It builds an AKID using the "match subject key identifier" option.

**Comment by mattw-db** [comment]
@bartonjs, thanks a lot!  We've already got a change in the works to not reuse CNs for test/local deployments.  It's good to hear that should be sufficient.  We'll look into the Authority Key Identifier and report back on that too.

Luckily, it seems like this issue would only be applicable to our test environments where alternate chains are used, so recovering from this is not a big deal. We can just issue new certs.

It might be nice to get a line-item added to the release notes and API documentation, since this was not the easiest of things to track down.
https://docs.microsoft.com/en-us/dotnet/core/whats-new/dotnet-core-3-0

Cheers,

Matt

**Comment by wfurt** [comment]
It seems like you get it working @mattw-db. Should this be closed or do we need documentation update? (or diagnostic improvements?)

**Comment by mattw-db** [comment]
@wfurt,

Yes, everything is working fine for us!  A note on the behavior change in the release notes may save other's some time, however.  I'd recommend that.  Otherwise, we can close this ticket.

Thanks,

Matt
